### PR TITLE
Omnibus Extension

### PR DESCRIPTION
**Pull request by arr00**
## Summary by Sourcery

Introduce an omnibus confidential transfer extension for ERC7984 and add corresponding mock, tests, and helpers.

New Features:
- Add ERC7984Omnibus extension with confidentialTransferOmnibus and confidentialTransferFromOmnibus functions and OmnibusTransfer event
- Expose omnibus extension in a mock contract ERC7984OmnibusMock for testing
- Add public $_mint helper in ERC7984Mock to simplify minting with uint64 inputs

Tests:
- Add comprehensive tests for omnibus transfers covering normal and over-balance scenarios for both transfer and transferFrom

Chores:
- Add changeset for the omnibus extension release

## Summary by CodeRabbit

- New Features
  - Added ERC7984Omnibus extension to enable confidential subaccount transfers via omnibus wallets, emitting OmnibusConfidentialTransfer events and enforcing encrypted-address access controls.

- Tests
  - Added comprehensive omnibus test suite covering transfer/from, callback, proof/no-proof permutations, balance edge cases, and permission checks; adjusted one test input value.

- Documentation
  - Updated token docs to include ERC7984Omnibus, its event, and usage guidance.

- Chores
  - Added package changeset and mock helpers/events for creating encrypted addresses and amounts; introduced omnibus-aware mock for testing.

### CODE REVIEW

File: test/token/ERC7984/extensions/ERC7984Omnibus.test.ts [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        await expect(
+          fhevm.userDecryptEaddress(omnibusTransferEvent.args[3], this.token.target, this.holder),
+        ).to.eventually.equal(this.subaccount.address);
+        await expect(
+          fhevm.userDecryptEuint(FhevmType.euint64, omnibusTransferEvent.args[4], this.token.target, this.holder),
+        ).to.eventually.equal(100);
+
+        await expect(this.acl.isAllowed(omnibusTransferEvent.args[2], this.holder)).to.eventually.be.true;
+        await expect(this.acl.isAllowed(omnibusTransferEvent.args[2], this.recipient)).to.eventually.be.true;
+        await expect(this.acl.isAllowed(omnibusTransferEvent.args[2], this.operator)).to.eventually.be.false;
```
[the remark this answers is not in the document]
Comment (arr00) [reply test/token/ERC7984/extensions/ERC7984Omnibus.test.ts:78]: Subaccount logic is all handled off-chain. The actual transfer logic is handled in the ERC7984 tests.

File: contracts/mocks/token/ERC7984OmnibusMock.sol [hunk 2]
```diff
@@ -0,0 +1,17 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.27;
+
+import {euint64} from "@fhevm/solidity/lib/FHE.sol";
+import {ERC7984Omnibus} from "../../token/ERC7984/extensions/ERC7984Omnibus.sol";
+import {ERC7984Mock, ERC7984} from "./ERC7984Mock.sol";
```
[the remark this answers is not in the document]
Comment (arr00) [reply contracts/mocks/token/ERC7984OmnibusMock.sol:7]: It does not fail compilation.

File: contracts/token/ERC7984/extensions/ERC7984Omnibus.sol [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        externalEaddress externalSender,
+        externalEaddress externalRecipient,
+        externalEuint64 externalAmount,
+        bytes calldata inputProof
+    ) public virtual returns (euint64) {
+        eaddress sender = FHE.fromExternal(externalSender, inputProof);
+        eaddress recipient = FHE.fromExternal(externalRecipient, inputProof);
+        euint64 amount = FHE.fromExternal(externalAmount, inputProof);
+
+        return confidentialTransferFromOmnibus(msg.sender, omnibusTo, sender, recipient, amount);
```
Comment (james-toussaint) [inline contracts/token/ERC7984/extensions/ERC7984Omnibus.sol:42]: Some allowances are given to `omnibusFrom` in `confidentialTransferFromOmnibus(..)` but it already has access to them.

File: contracts/token/ERC7984/extensions/ERC7984Omnibus.sol [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    }
+
+    function confidentialTransferOmnibus(
+        address omnibusTo,
+        eaddress sender,
+        eaddress recipient,
+        euint64 amount
+    ) public virtual returns (euint64) {
+        return confidentialTransferFromOmnibus(msg.sender, omnibusTo, sender, recipient, amount);
```
Comment (james-toussaint) [inline contracts/token/ERC7984/extensions/ERC7984Omnibus.sol:51]: For consistency it seems we should perform `FHE.isAllowed(..)` calls before.

File: contracts/token/ERC7984/extensions/ERC7984Omnibus.sol [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    ) public virtual returns (euint64) {
+        FHE.allowThis(sender);
+        FHE.allow(sender, omnibusFrom);
+        FHE.allow(sender, omnibusTo);
+
+        FHE.allowThis(recipient);
+        FHE.allow(recipient, omnibusFrom);
+        FHE.allow(recipient, omnibusTo);
+
+        euint64 transferred = confidentialTransferFrom(omnibusFrom, omnibusTo, amount);
```
Comment (james-toussaint) [inline contracts/token/ERC7984/extensions/ERC7984Omnibus.sol:84]: Why `transferred` isn't given access to other parties?

File: contracts/token/ERC7984/extensions/ERC7984Omnibus.sol [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        FHE.allow(recipient, omnibusTo);
+
+        FHE.allowThis(transferred);
+        FHE.allow(transferred, omnibusFrom);
+        FHE.allow(transferred, omnibusTo);
+
+        emit OmnibusConfidentialTransfer(omnibusFrom, omnibusTo, sender, recipient, transferred);
+        return transferred;
+    }
+}
```
Comment (james-toussaint) [inline contracts/token/ERC7984/extensions/ERC7984Omnibus.sol:155]: Inline docs are missing.

File: contracts/token/ERC7984/extensions/ERC7984Omnibus.sol (hunk 4 above)
Comment (arr00) [reply contracts/token/ERC7984/extensions/ERC7984Omnibus.sol:51]: Yes good catch

File: contracts/token/ERC7984/extensions/ERC7984Omnibus.sol [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+        externalEaddress externalSender,
+        externalEaddress externalRecipient,
+        externalEuint64 externalAmount,
+        bytes calldata inputProof
+    ) public virtual returns (euint64) {
+        eaddress sender = FHE.fromExternal(externalSender, inputProof);
+        eaddress recipient = FHE.fromExternal(externalRecipient, inputProof);
+        euint64 amount = FHE.fromExternal(externalAmount, inputProof);
+
+        return confidentialTransferFromOmnibus(omnibusFrom, omnibusTo, sender, recipient, amount);
```
[the remark this answers is not in the document]
Comment (arr00) [reply contracts/token/ERC7984/extensions/ERC7984Omnibus.sol:73]: Didn't do this since there is a lot of logic that would be duplicated by doing it naively. Could just make an internal function and have the specific functionality be part of the public funcs.

File: contracts/token/ERC7984/extensions/ERC7984Omnibus.sol [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+
+    function confidentialTransferFromOmnibus(
+        address omnibusFrom,
+        address omnibusTo,
+        eaddress sender,
+        eaddress recipient,
+        euint64 amount
+    ) public virtual returns (euint64) {
+        require(FHE.isAllowed(sender, msg.sender), ERC7984UnauthorizedUseOfEncryptedAddress(sender));
+        require(FHE.isAllowed(recipient, msg.sender), ERC7984UnauthorizedUseOfEncryptedAddress(recipient));
```
Comment (james-toussaint) [inline contracts/token/ERC7984/extensions/ERC7984Omnibus.sol:92]: No need to:
```suggestion
        require(FHE.isAllowed(recipient, msg.sender), ERC7984UnauthorizedUseOfEncryptedAddress(recipient));
        require(FHE.isAllowed(amount, msg.sender), ERC7984UnauthorizedUseOfEncryptedAmount(amount, msg.sender));
```
?
And update other event to `ERC7984UnauthorizedUseOfEncryptedAddress(euint64 addr, address user)`?
